### PR TITLE
Feature/domain

### PR DESCRIPTION
**Pull request by whaeck**
This prepares for the base class of a function object by introducing domain objects. These objects will simplify domain testing in function objects.

The domain of definition or simply the domain of a function is the set of input or argument values for which the function y -> f(x) is defined.

The following domains are currently available:
- open domain: all values of x are allowed
- interval domain: all values of x in an interval [a,b] are allowed

### CODE REVIEW

File: src/scion/math/IntervalDomain/src/verifyDomain.hpp [hunk 1]
```diff
@@ -0,0 +1,10 @@
+static void verifyDomain( const X& lower, const X& upper ) {
+
+  if ( ( upper < lower ) || ( upper == lower ) ) {
+
+    Log::error( "Incoherent domain, expected lower > upper" );
```
Comment (BobbiRiedel) [inline src/scion/math/IntervalDomain/src/verifyDomain.hpp:5]: Fix typo; upper > lower

File: python/test/math/Test_scion_math_IntervalDomain.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+            self.assertEqual( False, chunk.is_inside( x = min ) );
+            self.assertEqual( True, chunk.is_inside( x = -1.0 ) );
+            self.assertEqual( True, chunk.is_inside( x = -0.5 ) );
+            self.assertEqual( True, chunk.is_inside( x =  0.0 ) );
+            self.assertEqual( True, chunk.is_inside( x =  0.5 ) );
+            self.assertEqual( True, chunk.is_inside( x =  1.0 ) );
+            self.assertEqual( False, chunk.is_inside( x = max ) );
+
+            self.assertEqual( False, chunk.is_inside( x = min ) );
```
Comment (BobbiRiedel) [inline python/test/math/Test_scion_math_IntervalDomain.py:33]: is_contained should be used instead

File: python/test/math/Test_scion_math_OpenDomain.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+            self.assertEqual( True, chunk.is_inside( x = min ) );
+            self.assertEqual( True, chunk.is_inside( x = -1.0 ) );
+            self.assertEqual( True, chunk.is_inside( x = -0.5 ) );
+            self.assertEqual( True, chunk.is_inside( x =  0.0 ) );
+            self.assertEqual( True, chunk.is_inside( x =  0.5 ) );
+            self.assertEqual( True, chunk.is_inside( x =  1.0 ) );
+            self.assertEqual( True, chunk.is_inside( x = max ) );
+
+            self.assertEqual( True, chunk.is_inside( x = min ) );
```
Comment (BobbiRiedel) [inline python/test/math/Test_scion_math_OpenDomain.py:29]: is_inside should be is_contained

File: src/scion/math/IntervalDomain/test/IntervalDomain.test.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+        CHECK( false == chunk.isInside( min ) );
+        CHECK( true == chunk.isInside( -1.0 ) );
+        CHECK( true == chunk.isInside( -0.5 ) );
+        CHECK( true == chunk.isInside(  0.0 ) );
+        CHECK( true == chunk.isInside(  0.5 ) );
+        CHECK( true == chunk.isInside(  1.0 ) );
+        CHECK( false == chunk.isInside( max ) );
+
+        CHECK( false == chunk.isInside( min ) );
```
Comment (BobbiRiedel) [inline src/scion/math/IntervalDomain/test/IntervalDomain.test.cpp:39]: is_inside should be is_contained

File: src/scion/math/OpenDomain/test/OpenDomain.test.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+        CHECK( true == chunk.isInside( min ) );
+        CHECK( true == chunk.isInside( -1.0 ) );
+        CHECK( true == chunk.isInside( -0.5 ) );
+        CHECK( true == chunk.isInside(  0.0 ) );
+        CHECK( true == chunk.isInside(  0.5 ) );
+        CHECK( true == chunk.isInside(  1.0 ) );
+        CHECK( true == chunk.isInside( max ) );
+
+        CHECK( true == chunk.isInside( min ) );
```
Comment (BobbiRiedel) [inline src/scion/math/OpenDomain/test/OpenDomain.test.cpp:33]: is_inside should be is_contained

**Review by BobbiRiedel — APPROVED** [review]

File: src/scion/math/IntervalDomain/src/verifyDomain.hpp (hunk 1 above)
Comment (whaeck) [reply src/scion/math/IntervalDomain/src/verifyDomain.hpp:5]: Consider it fixed, see #45

File: python/test/math/Test_scion_math_IntervalDomain.py (hunk 2 above)
Comment (whaeck) [reply python/test/math/Test_scion_math_IntervalDomain.py:33]: Yes, I noticed this last week and fixed it in #42

File: src/scion/math/OpenDomain/test/OpenDomain.test.cpp (hunk 5 above)
Comment (whaeck) [reply src/scion/math/OpenDomain/test/OpenDomain.test.cpp:33]: I actually forgot this one.

**Pull request merged**